prometheus/metrics.go

```diff
@@ -129,40 +131,65 @@ func getServiceMetricsAsync(api v1.API, namespace string, servicename string, du
 		Histograms: histograms}
 }
 
-func fetchRateRange(api v1.API, metricName string, labels string, bounds v1.Range, rateInterval string, ch chan Metric) {
-	query := fmt.Sprintf("rate(%s%s[%s])", metricName, labels, rateInterval)
+func joinLabels(labels []string) string {
+	str := ""
+	if len(labels) > 0 {
+		sep := ""
+		for _, lbl := range labels {
+			str = str + sep + lbl
+			sep = ","
+		}
+	}
+	return str
+}
+
+func fetchRateRange(api v1.API, metricName string, labels string, grouping string, bounds v1.Range, rateInterval string, ch chan Metric) {
+	var query string
+	if grouping == "" {
+		query = fmt.Sprintf("round(sum(irate(%s%s[%s])), 0.001)", metricName, labels, rateInterval)
+	} else {
+		query = fmt.Sprintf("round(sum(irate(%s%s[%s])) by (%s), 0.001)", metricName, labels, rateInterval, grouping)
+	}
 	fetchRange(api, query, bounds, ch)
 }
 
-func fetchHistogramRange(api v1.API, metricName string, labels string, bounds v1.Range, rateInterval string, ch chan Histogram) {
+func fetchHistogramRange(api v1.API, metricName string, labels string, grouping string, bounds v1.Range, rateInterval string, ch chan Histogram) {
 	// Note: we may want to make returned stats configurable in the future
 	avgChan, medChan, p95Chan, p99Chan := make(chan Metric), make(chan Metric), make(chan Metric), make(chan Metric)
 
+	groupingAvg := ""
+	groupingQuantile := ""
+	if grouping != "" {
+		groupingAvg = fmt.Sprintf(" by (%s)", grouping)
+		groupingQuantile = fmt.Sprintf(",%s", grouping)
+	}
+
 	// Average
 	go func() {
 		query := fmt.Sprintf(
-			"sum(rate(%s_sum%s[%s])) / sum(rate(%s_count%s[%s]))", metricName, labels, rateInterval, metricName, labels, rateInterval)
+			"sum(rate(%s_sum%s[%s]))%s / sum(rate(%s_count%s[%s]))%s", metricName, labels, rateInterval, groupingAvg,
+			metricName, labels, rateInterval, groupingAvg)
 		fetchRange(api, query, bounds, avgChan)
 	}()
 
 	// Median
 	go func() {
 		query := fmt.Sprintf(
-			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, rateInterval)
+			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le%s))", metricName, labels, rateInterval, groupingQuantile)
```

Comment:
I'm not familiar with syntax of p8s on this one.
"le%s" is the go syntax here ?
(first thought it was a typo)

Reply:
`groupingQuantile` is a string built above in this form: `,foo,bar` so it translates in `sum (...) by (le,foo,bar)` which is what I want

Reply:
ok, it was hard to follow, as I wasn't sure if the le was intentional.

Reply:
Minor, if you can add a comment, perhaps it would clarify to others, just the same thing you put in the PR

Reply:
yeah, these strings builder are not the prettiest thing... I'd wish we had a nice DSL to query prometheus.

Reply:
@lucasponce done I added 2 examples